source/common/grpc/json_transcoder_filter.cc

```diff
@@ -0,0 +1,398 @@
+  std::string path = headers.Path()->value().c_str();
+  std::string args;
+
+  const size_t pos = path.find('?');
+  if (pos != std::string::npos) {
+    args = path.substr(pos + 1);
+    path = path.substr(0, pos);
+  }
+
+  RequestInfo request_info;
+  std::vector<VariableBinding> variable_bindings;
+  method_descriptor =
+      path_matcher_->Lookup(method, path, args, &variable_bindings, &request_info.body_field_path);
+  if (!method_descriptor) {
+    return Status(Code::NOT_FOUND, "Could not resolve " + path + " to a method");
+  }
+
+  Status status = methodToRequestInfo(method_descriptor, &request_info);
+  if (!status.ok()) {
+    return status;
+  }
+
+  for (const auto& binding : variable_bindings) {
+    google::grpc::transcoding::RequestWeaver::BindingInfo resolved_binding;
+    status = type_helper_->ResolveFieldPath(*request_info.message_type, binding.field_path,
+                                            &resolved_binding.field_path);
+    if (!status.ok()) {
+      return status;
+    }
+
+    resolved_binding.value = binding.value;
+
+    request_info.variable_bindings.emplace_back(std::move(resolved_binding));
+  }
+
+  std::unique_ptr<JsonRequestTranslator> request_translator{
+      new JsonRequestTranslator(type_helper_->Resolver(), &request_input, request_info,
+                                method_descriptor->client_streaming(), true)};
+
+  const auto response_type_url =
+      TYPE_URL_PREFIX + "/" + method_descriptor->output_type()->full_name();
+  std::unique_ptr<ResponseToJsonTranslator> response_translator{
+      new ResponseToJsonTranslator(type_helper_->Resolver(), response_type_url,
+                                   method_descriptor->server_streaming(), &response_input)};
+
+  transcoder.reset(
+      new TranscoderImpl(std::move(request_translator), std::move(response_translator)));
+  return Status::OK;
+}
+
+Status JsonTranscoderConfig::methodToRequestInfo(const google::protobuf::MethodDescriptor* method,
+                                                 google::grpc::transcoding::RequestInfo* info) {
+  auto request_type_url = TYPE_URL_PREFIX + "/" + method->input_type()->full_name();
+  info->message_type = type_helper_->Info()->GetTypeByTypeUrl(request_type_url);
+  if (info->message_type == nullptr) {
+    ENVOY_LOG(debug, "Cannot resolve input-type: {}", method->input_type()->full_name());
+    return Status(Code::NOT_FOUND, "Could not resolve type: " + method->input_type()->full_name());
+  }
+
+  return Status::OK;
+}
+
+JsonTranscoderFilter::JsonTranscoderFilter(JsonTranscoderConfig& config) : config_(config) {}
+
+Http::FilterHeadersStatus JsonTranscoderFilter::decodeHeaders(Http::HeaderMap& headers,
+                                                              bool end_stream) {
+  const auto status =
+      config_.createTranscoder(headers, request_in_, response_in_, transcoder_, method_);
+
+  if (!status.ok()) {
+    return Http::FilterHeadersStatus::Continue;
```

Comment:
Add a comment, we shouldn't fail the request because the transcoder creation failed. The request might be a normal gRPC request that it should be passed to upstream.